### PR TITLE
En/review eg/fields surfaces objective functions mod

### PR DESCRIPTION
**Pull request by eduardolneto**
Small changes on fields.py class near_axis to work as a dependecy of custom_losses. Refactoring loss functions in objective_functions.py to meet the standard of new custom losses implementation. Adding surface-coil disstance, linking number and coils forces loss functions. Mostly changes to fields.py and objective_functions.py

### CODE REVIEW

File: essos/surfaces.py
Comment (EstevaoMGomes) [inline essos/surfaces.py:1]: Remove these changes. Removing spaces is fine, but it may cause some merge issues with other people's PRs.

File: essos/fields.py [hunk 1]
```diff
@@ -424,7 +424,8 @@ def dofs(self):
     
     @dofs.setter
     def dofs(self, new_dofs):
-        self._dofs = jnp.array(new_dofs)
+        # Ensure dofs are always float for JAX autodiff compatibility
+        self._dofs = jnp.array(new_dofs, dtype=float)
```
Comment (EstevaoMGomes) [inline essos/fields.py:428]: This should not be needed...

File: essos/objective_functions.py
Comment (EstevaoMGomes) [inline essos/objective_functions.py:1]: I like that this was simplified, but I still need to check the losses more thoroughly. An example where they are used would be useful for that.

Comment (eduardolneto) [reply essos/objective_functions.py:1]: @EstevaoMGomes  In the examples, I kept similar local loss functions because that was the exisiting structure there. I could change them to always improt from this file. However, adding those examples to this PR would  imply that I would need to add pleanty of file modifications for the examples. Which is something I thought you did not want previously.

File: essos/fields.py (hunk 1 above)
Comment (eduardolneto) [reply essos/fields.py:428]: This was related with the way near axis is constructed. Some values are technically integers at the near-axis class level. This may not be needed now. However, if we are going to use a jax pyqsc version that may actually be necessary again because I think the original version may output some of the quantities as integers. I would keep it for now and change it later.

File: essos/surfaces.py
Comment (eduardolneto) [reply essos/surfaces.py:1]: I think this is fine. The space was removed and added again, so there are virtually no chnages? This also looks the same as the file in eg/analysis branch to me.

**Review by rogeriojorge — CHANGES_REQUESTED** [review]
1. objective_functions.py:222, 234, and 243: loss_BdotN(), loss_BdotN_only(), and loss_BdotN_only_constraint() still call field_from_dofs(...), but this PR removes field_from_dofs() earlier in the file. These functions will now raise NameError. You could make a small change of these losses to accept a field directly like the new custom-loss style.

2. I didn't test locally, but seems that objective_functions.py:81, 92, 114, 135, and 137: particle-loss functions use xyz[:,0], xyz[:,1], xyz[:,2], which selects timesteps, not coordinates, since xyz has shape (n_particles, n_times, 3). Should this be xyz[:, :, 0], xyz[:, :, 1], xyz[:, :, 2]? Otherwise jnp.diff(..., axis=1) is not differentiating the intended time series.

3. objective_functions.py:155 and 170: loss_particle_Br() and loss_particle_iota() accept a timestep argument but still pass timestep=1.e-8 to Tracing. I’d replace those with timestep=timestep, otherwise the user-provided value is ignored.

Minor things

objective_functions.py:460-505, loss_lorentz_force_coils(): block_size is in the signature, static args, and docstring, but it is not used. I’d either remove it from this function for now, or actually use it to block the mutual-field calculation. Also n_points on line 465 is unused.

new_nearaxis_from_x_and_old_nearaxis and other imports like lax might be unused and could be deleted

File: essos/fields.py (hunk 1 above)
Comment (EstevaoMGomes) [reply essos/fields.py:428]: Then I think it should be changed in JAX PyQSC and not here. But it's ok to keep now.

**Comment by eduardolneto** [comment]
I have addressed all comments. Added an exampel to test surface-distance distance-distance, linkingnumber and coisl forces. There are some incositencies with coils.py. However, these shjould only be addressed in a PR once all the current PR #40 #38 #39 #37 #36  are merged. Because it will need to use the updated coils.py file.

File: essos/objective_functions.py [hunk 2]
```diff
@@ -243,44 +163,74 @@ def loss_iota(x,particles,dofs_curves, currents_scale, nfp,n_segments=60, stells
...
-# @partial(jit, static_argnums=(0, 1))
-def loss_normB_axis_average(x,dofs_curves,currents_scale,nfp,n_segments=60,stellsym=True, npoints=15,target_B_on_axis=5.7):
-    field=field_from_dofs(x,dofs_curves,currents_scale,nfp,n_segments,stellsym)
-    R_axis=field.r_axis
-    phi_array = jnp.linspace(0, 2 * jnp.pi, npoints)
-    B_axis = vmap(lambda phi: field.AbsB(jnp.array([R_axis * jnp.cos(phi), R_axis * jnp.sin(phi), 0])))(phi_array)
-    return jnp.array([jnp.absolute(jnp.average(B_axis)-target_B_on_axis)])
+
+def BdotN_constraint(field,surface):
+    return jnp.sum(jnp.abs(BdotN_over_B(surface, field)))
```
Comment (rogeriojorge) [inline essos/objective_functions.py]: This is defining BdotN_constraint twice, seems like the first one can be deleted

File: essos/objective_functions.py [hunk 3]
```diff
@@ -243,44 +163,74 @@ def loss_iota(x,particles,dofs_curves, currents_scale, nfp,n_segments=60, stells
...
+    bdotn_over_b = BdotN_over_B(surface, field)
+    bdotn_over_b_loss = jnp.sqrt(jnp.sum(jnp.maximum(jnp.square(bdotn_over_b)-target_tol,0.0)))
+    return bdotn_over_b_loss
+
+
+###########################  B ON SURAFCE LOSSES FOR STOCHASTIC OPTIMIZATION ##########################
+def copy_coils_from_field(field):
+    return field.coils.copy()
+
+@partial(jit, static_argnames=['key', 'sampler'])
```
Comment (rogeriojorge) [inline essos/objective_functions.py]: Can key / keys from perturbed_field_from_field() and the stochastic losses be marked as dynamic instead of static? That may avoid recompiles. I think sampler is ok to be static

File: examples/coil_optimization/optimize_coils_vmec_surface.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 #  `scipy.optimize.minimize` or `jaxopt`, can be used as well and may even be preferable.
 from scipy.optimize import least_squares
 
-input_filepath = os.path.join(os.path.dirname(__file__), "..", "input_files")
+input_filepath = os.path.join(os.path.dirname(__name__), "input_files")
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface.py]: Isn't os.path.dirname(__file__) less prone to error? And given that this is inside coil optimization, shouldn't we keep the previous ".."?

File: essos/surfaces.py
Comment (rogeriojorge) [reply essos/surfaces.py:1]: I think I agree with @EstevaoMGomes here, less noise is better!

**Review by rogeriojorge — CHANGES_REQUESTED** [review]
Only small 4 changes to make before merging

File: essos/surfaces.py
Comment (eduardolneto) [reply essos/surfaces.py:1]: Again this was removed and added, if you compare the files in both original branch and this PR they are identical.

File: essos/objective_functions.py (hunk 2 above)
Comment (eduardolneto) [reply essos/objective_functions.py]: That is supposed to not have the constraint, it is the loss version, I have removed the double constraint.

File: essos/objective_functions.py (hunk 3 above)
Comment (eduardolneto) [reply essos/objective_functions.py]: Done

**Comment by eduardolneto** [comment]
Addressed remaining comments @rogeriojorge @EstevaoMGomes

**Review by EstevaoMGomes — APPROVED** [review]
Looks good to be merged after [PR!38](https://github.com/uwplasma/ESSOS/pull/38), [PR!39](https://github.com/uwplasma/ESSOS/pull/39), and [PR!34](https://github.com/uwplasma/ESSOS/pull/34).

**Review by rogeriojorge — APPROVED** [review]
Looks good to me
